sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/SortDirection.java

```diff
@@ -19,14 +19,33 @@
 
 import org.apache.spark.annotation.Experimental;
 
+import static org.apache.spark.sql.connector.expressions.NullOrdering.NULLS_FIRST;
+import static org.apache.spark.sql.connector.expressions.NullOrdering.NULLS_LAST;
+
 /**
  * A sort direction used in sorting expressions.
+ * <p>
+ * Each direction has a default null ordering that is implied if no null ordering is specified
+ * explicitly.
  *
  * @since 3.2.0
  */
 @Experimental
 public enum SortDirection {
-  ASCENDING, DESCENDING;
+  ASCENDING(NULLS_FIRST), DESCENDING(NULLS_LAST);
+
+  private final NullOrdering defaultNullOrdering;
+
+  SortDirection(NullOrdering defaultNullOrdering) {
+    this.defaultNullOrdering = defaultNullOrdering;
+  }
+
+  /**
+   * Returns the default null ordering to use if no null ordering is specified explicitly.
+   */
+  public NullOrdering defaultNullOrdering() {
```

Comment:
This makes sense, because the catalyst `SortDirection` also has a default null ordering.

Shall we further follow catalyst and add an overload of `Expressions.sort` which doesn't take `NullOrdering` and infer null ordering from `SortDirection`?

Reply:
Just 2 cents, I'd feel more natural to change the null ordering "optional" in SortOrder. It'd be more intuitive on the behavior we want, "if null ordering is not presented, default null ordering of SortDirection is taken". Callers of SortOrder has to resolve the null ordering but then we no longer delegate the implementations to do that instead.

Each implementation can still simply call the method to get the default null ordering and provide it so no strong voice. Just would like to think out loud what'd be the clearer.

Reply:
If we decide to introduce this, I think it makes sense to leverage it in `Expressions$sort` as @cloud-fan said. That's a good use case.

@HeartSaVioR, could you give an example, please? I am not sure I fully got it. Do you mean to make `nullOrdering` in `SortOrder` optional instead of exposing a default null ordering in `SortDirection`? If so, I usually prefer to use the simplest types in the API. Using `Optional` will make the API harder to use, in my view. In addition, we probably want to handle `SortDirection` in the same way as it is also optional.

Reply:
So the current implementation of SortOrder is following:

```
/**
 * Represents a sort order in the public expression API.
 *
 * @since 3.2.0
 */
@Experimental
public interface SortOrder extends Expression {
  /**
   * Returns the sort expression.
   */
  Expression expression();

  /**
   * Returns the sort direction.
   */
  SortDirection direction();

  /**
   * Returns the null ordering.
   */
  NullOrdering nullOrdering();
}
```

which we force to implement `nullOrdering()` which destroys the purpose of having default null ordering in SortDirection.

Given SortDirection will have default null ordering, we can provide "default implementation" of `nullOrdering()` which would be simply `return direction().defaultNullOrdering();` and add some explanation to javadoc to override this method if they don't want to follow the default null ordering of SortDirection.

(I actually thought about Optional and I found by myself not prefer it. Given we're able to leverage default method in interface, I don't see the reason not to use it.)

Reply:
As I said as "2 cents", no strong opinion on this (implementors just need to add the line by themselves, which doesn't seem to be critical) and I'll leave the decision to you.

Reply:
@HeartSaVioR `SortOrder` is not an API we ask users to implement. The users should call `Expressions.sort` to create an instance of `SortOrder`.

That's why I suggested adding an overload of `Expressions.sort`, as that makes null ordering optional to the users.

Reply:
Ah OK. I was misunderstood about the usage on SortOrder. If the users (I assume this includes data source implementation) are discouraged to implement this, no problem at all and please disregard my comments.

I understand the catalyst module is considered as private API, but given connector package is now playing as public API (at least some of APIs are public API data source implementations should leverage), it'd be nice if we could explicitly separate the public vs private APIs.